jgrapht-core/src/main/java/org/jgrapht/alg/cycle/BergeGraphInspector.java

```diff
@@ -0,0 +1,1106 @@
+        Set<V> temp = new HashSet<V>();
+        temp.addAll(Y(g,Nab,c));
+        temp.addAll(W(g,Nab,c));
+        Set<V> res = new HashSet<V>();
+        for (V it : g.vertexSet()){
+            if (isYXComplete(g, it, temp))
+                res.add(it);
+        }
+        return res;
+    }
+    
+    /**
+     * X(a,b,c)=Y(a,b,c)+Z(a,b,c)
+     * @param g A graph
+     * @param Nab The set of all {a,b}-complete vertices
+     * @param c A vertex
+     * @return The union of Y(a,b,c) and Z(a,b,c)
+     */
+    private Set<V> X(Graph<V,E> g, Set<V> Nab, V c){
+        Set<V> res = new HashSet<V>();
+        res.addAll(Y(g,Nab,c));
+        res.addAll(Z(g,Nab,c));
+        return res;
+    }
+    
+    /**
+     * A triple (a,b,c) of vertices is relevant if a,b are distinct and nonadjacent, and c is not contained in N(a,b) (possibly
+     * c is contained in {a,b}).
+     * @param g A graph
+     * @param a A vertex
+     * @param b A vertex
+     * @param c A vertex
+     * @return Assessement whether a,b,c is a relevant triple
+     */
+    private boolean isTripleRelevant(Graph<V,E> g, V a, V b, V c){
+        return a!=b&&!g.containsEdge(a,b)&&!N(g,a,b).contains(c);
+    }
+    
+    
+    /**
+     * Returns a set of vertex sets that may be near-cleaners for an amenable hole in g.
+     * @param g A graph
+     * @return possible near-cleaners
+     */
+    Set<Set<V>> routine3(Graph<V,E> g){
+        Set<Set<V>> NuvList = new HashSet<Set<V>>();
+        for (V u : g.vertexSet()){
+            for (V v : g.vertexSet()){
+                if (u==v||!g.containsEdge(u,v)) continue;
+                NuvList.add(N(g,u,v));
+            }
+        }
+        
+        Set<Set<V>> tripleList = new HashSet<Set<V>>();
+        for (V a : g.vertexSet()){
+            for (V b : g.vertexSet()){
+                if (a==b||g.containsEdge(a,b)) continue;
+                Set<V> Nab = N(g,a,b);
+                for (V c : g.vertexSet()){
+                    if (isTripleRelevant(g,a,b,c)){
+                        tripleList.add(X(g,Nab,c));
+                    }
+                }
+            }
+        }
+        Set<Set<V>> res = new HashSet<Set<V>>();
+        for (Set<V> Nuv : NuvList){
+            for (Set<V> triple : tripleList){
+                Set<V> temp = new HashSet<V>();
+                temp.addAll(Nuv);
+                temp.addAll(triple);
+                res.add(temp);
+            }
+        }
+        return res;
+    }
+    
+    /**
+     * Performs the Berge Recognition Algorithm.
+     * <p> First this algorithm is used to test whether g or its complement contain a jewel, a pyramid or a configuration of type
+     * 1, 2 or 3. If so, it is output that g is not Berge. If not, then every shortest odd hole in g is amenable. This asserted, the near-cleaner subsets
+     * of V(g) are determined. For each of them in turn it is checked, if this subset is a near-cleaner and, thus, if there is an odd hole. If 
```

Comment:
hm why did you revert back the math formatting? E.g. V(g)-> $V(G)$, g->$G$, 

Running this method takes O(|V|^9), and computing the certificate takes O(|V|^5). -> Running this method takes $O(|V|^9)$, and computing the certificate takes an additional $O(|V|^5)$.

Reply:
Sorry, I changed it back.